graph/src/data/store/mod.rs

```diff
@@ -439,6 +440,75 @@ impl<'a> From<Vec<(&'a str, Value)>> for Entity {
     }
 }
 
+pub trait TryFromEntity: Sized {
+    const ENTITY_TYPE: &'static str;
+
+    fn attribute_missing(attr: impl Into<String>) -> Error {
+        format_err!(
+            "{} attribute \"{}\" is missing",
+            Self::ENTITY_TYPE,
+            attr.into(),
+        )
+    }
+
+    fn attribute_invalid(
+        attr: impl Into<String>,
+        expected_type: impl Into<String>,
+        value: &Value,
+    ) -> Error {
+        format_err!(
+            "{} attribute \"{}\" is not a {}: {:?}",
+            Self::ENTITY_TYPE,
+            attr.into(),
+            expected_type.into(),
+            value
+        )
+    }
+
+    fn extract_optional<T, F>(
+        entity: &mut Entity,
+        name: impl Into<String>,
+        type_name: impl Into<String>,
+        f: F,
+    ) -> Result<Option<T>, Error>
+    where
+        F: FnOnce(Value) -> Option<T>,
+        T: Sized,
+    {
+        let name = name.into();
+        match entity.remove(&name) {
+            Some(value) => match f(value.clone()) {
+                Some(converted_value) => Ok(Some(converted_value)),
+                None => Err(Self::attribute_invalid(name, type_name, &value)),
+            },
+            None => Ok(None),
+        }
+    }
+
+    fn extract<T, F>(
+        entity: &mut Entity,
+        name: impl Into<String>,
+        type_name: impl Into<String>,
+        f: F,
```

Comment:
We have a `ValueType` enum in the graph crate -- how about using that in place of `type_name`? Then this method could also take care of calling `.as_string()`, etc

Reply:
Yes. I didn't like the type name as a string. Let's do it.